text/1444-union.md

```diff
@@ -178,39 +190,489 @@ fn is_zero(v: Value) -> bool {
+Note that those maximums may come from different fields; for instance:
+
+```rust
+#[repr(C)]
 union U {
-    f1: u32,
-    f2: f32,
+    f1: u16,
+    f2: [u8; 4],
+}
+
+fn main() {
+    assert_eq!(std::mem::size_of<U>(), 4);
+    assert_eq!(std::mem::align_of<U>(), 2);
 }
+```
+
+Alignment of a union with `#[repr(C, packed)]` is typically equal to 1, and its
+size is equal to the size of its largest field (so packed unions do not have
+trailing padding).
+
+By default, if `#[repr(C)]` is not used, layout of a union is *unspecified*.
+
+Regardless of representation attributes, all the union fields and the union
+itself are located at the same address in memory (pointers to them are equal).  
+For `#[repr(C)]` unions this property is guaranteed by ISO C standards.
+
+### Dynamically sized fields
+
+Unions cannot contain fields with dynamically sized types like `[u8]` or
+`Trait`.
 
-#[test]
+### Unions and `#[derive]`
+
+Unions support deriving a few traits that do not require accessing their fields.
+Supported traits are
+
+- `Copy`, supported only for unions with `Copy` fields
+- `Clone`, supported only for unions implementing `Copy`, which can be cloned
+trivially
+- `Eq` (but not `PartialEq`), supported for unions with `Eq` fields.
+
+Since accessing union fields reliably requires extra knowledge, traits trying to
+do it (e.g. `PartialEq`) cannot be derived automatically.
+
+## Constructing union values
+
+A value of a union type can be created using the same syntax that is used for
+struct types, except that it must specify exactly one field:
+
+```rust
+let u = MyUnion { f1: 1 };
+```
+
+This expression will create a value of type `MyUnion` with active field `f1`.  
+Union expressions, unlike struct expressions, do not support functional record
+update (FRU) `MyUnion { f1: 1, ..v }`.  
+Note that creating a union with union expression is safe (no potentially
+unreliable data is accessed) and therefore doesn't require `unsafe` block.
+
+Alternatively, a union can be created in uninitialized state and filled in
+later.
+```rust
+let u: MyUnion;
+u.f1 = 1; // u becomes initialized, f1 is the active field
+          // NB: this is not actually implemented right now, but it should be
+```
+
+## Accessing union fields
+
+### Field access (`.`)
+
+Union fields can be accessed using the same dot syntax as struct fields
+```rust
+unsafe {
+    let a = u.f1;
+    u.f2 = 1.0;
+    fn_call(&u.f2);
+}
+```
+
+All accesses to union fields (both reads and writes, and also borrows) require
+`unsafe` blocks.
+The active field/fragment of a union can always be accessed safely, but
+reads of inactive fields/fragments can result in undefined behavior (see
+[Guarantees and undefined behavior](#guarantees-and-undefined-behavior) for
+more detail).
+
+Write to a union field can potentially overwrite contents of its other fields.  
+Write to a union field should not modify any bits outside of the modified field
```

Comment:
To support [this](https://github.com/petrochenkov/rfcs/blob/8f1a960844b6ae37ec19fd89d39355fa98b1bb2b/text/1444-union.md#history-of-assignments).
(This guarantee was in the original RFC as well.)